lang/c++/impl/NodeImpl.cc

```diff
@@ -201,14 +205,183 @@ NodeRecord::printJson(std::ostream &os, int depth) const
+      for (unsigned int i = 0; i < vg.size(); i++) {
+        std::string hex_string = int_to_hex(static_cast<int>(vg[i]));
+        for (int j = 0; j < kByteStringSize; j++) {
+          s[i*kByteStringSize + j] = hex_string[j];
+        }
+      }
+      os << "\"" << s << "\"";
+    } break;
+    default:
+      break;
+  }
+}
+
+void NodeEnum::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                  int depth) const {
+  assert(g.type() == AVRO_ENUM);
+  os << "\"" << g.value<GenericEnum>().symbol() << "\"";
+}
+
+void NodeFixed::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                   int depth) const {
+  assert(g.type() == AVRO_FIXED);
+  // ex: "\uOOff"
+  // Convert to a string
+  const std::vector<uint8_t> &vg = g.value<GenericFixed>().value();
+  std::string s;
+  s.resize(vg.size() * kByteStringSize);
+  for (unsigned int i = 0; i < vg.size(); i++) {
+    std::string hex_string = int_to_hex(static_cast<int>(vg[i]));
+    for (int j = 0; j < kByteStringSize; j++) {
+      s[i*kByteStringSize + j] = hex_string[j];
+    }
+  }
+  os << "\"" << s << "\"";
+}
+
+void NodeUnion::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                   int depth) const {
+  assert(g.type() == AVRO_UNION);
+  // ex: "type": [ "string", "int" ], "default": "sval"
+  if (g.value<GenericUnion>().datum().type() == AVRO_NULL) {
+    os << "null";
+  } else {
+    leafAt(0)->printDefaultToJson( g.value<GenericUnion>().datum(), os, depth);
+  }
+}
+
+void NodeArray::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                   int depth) const {
+  assert(g.type() == AVRO_ARRAY);
+  // ex: "default": [1]
+  if (g.value<GenericArray>().value().empty()) {
+    os << "[]";
+  } else {
+    os << "[\n";
+    depth++;
+
+    // Serialize all values of the array with recursive calls:
+    for (unsigned int i = 0; i < g.value<GenericArray>().value().size(); i++) {
+      if (i > 0) {
+        os << ",\n";
+      }
+      os << indent(depth);
+      leafAt(0)->printDefaultToJson(g.value<GenericArray>().value()[i], os,
+                                    depth);
+    }
+    os << "\n" << indent(--depth) << "]";
+  }
+}
+
+void NodeSymbolic::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                      int depth) const {
+  getNode()->printDefaultToJson(g, os, depth);
+}
+
+void NodeRecord::printDefaultToJson(const GenericDatum &g, std::ostream &os,
+                                    int depth) const {
+  assert(g.type() == AVRO_RECORD);
+  if (g.value<GenericRecord>().fieldCount() == 0) {
+    os << "{}";
+  } else {
+    os << "{\n";
+
+    // Serialize all fields of the record with recursive calls:
+    for (unsigned int i = 0; i < g.value<GenericRecord>().fieldCount(); i++) {
+      if (i == 0) {
+        ++depth;
+      } else {  // i > 0
+        os << ",\n";
```

Comment:
I suggest putting the newlines before the indent and value to simplify the formatting a bit

Reply:
I've updated the PR quite a bit. Please take a look - we've been using this code/feature successfully now for many months at Google BigQuery and would like to have this officially added to the open source codebase.